pkg/roachprod/vm/ibm/helpers.go

```diff
@@ -0,0 +1,1205 @@
+			vpcService.NewGetNetworkInterfaceFloatingIPOptions(
+				networkInterfaceID,
+				floatingIPAddressID,
+			),
+		)
+		if err != nil {
+			l.Printf("	=> failed to get floating IP address %s for instance %s: %s", floatingIPAddressID, instanceID, err)
+			return p.attachOrCreateFloatingIPAddress(
+				l,
+				zone,
+				resourceGroupID,
+				instanceID,
+				networkInterfaceID,
+				attempt+1,
+			)
+		}
+
+		return &vpcv1.FloatingIP{
+			Address: res.Address,
+			ID:      res.ID,
+			CRN:     res.CRN,
+			Name:    res.Name,
+		}, nil
+	}
+
+	l.Printf(
+		"    -> no unbound IP address found, creating a new one for instance %s...",
+		instanceID,
+	)
+	// Create a new floating IP address
+	res, _, err := vpcService.CreateFloatingIP(
+		vpcService.NewCreateFloatingIPOptions(
+			&vpcv1.FloatingIPPrototype{
+				Target: &vpcv1.FloatingIPTargetPrototype{
+					ID: &networkInterfaceID,
+				},
+				ResourceGroup: &vpcv1.ResourceGroupIdentity{
+					ID: &p.config.roachprodResourceGroupID,
+				},
+			},
+		),
+	)
+	if err != nil {
+		return nil, errors.Wrapf(
+			err,
+			"failed to create floating IP for instance %s in zone %s",
+			instanceID,
+			zone,
+		)
+	}
+
+	return res, nil
+}
+
+// getImageIDForRegion retrieves the image ID for a given image name and region.
+func (p *Provider) getImageIDForRegion(l *logger.Logger, imageName, region string) (string, error) {
+
+	p.cacheMutex.Lock()
+	defer p.cacheMutex.Unlock()
+
+	if p.cachedImageIDs[region][imageName] != "" {
+		return p.cachedImageIDs[region][imageName], nil
+	}
+
+	vpcService, err := p.getVpcService(region)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to get VPC service while querying image ID")
+	}
+
+	imagesPager, err := vpcService.NewImagesPager(
+		vpcService.NewListImagesOptions().SetName(imageName).SetLimit(defaultPaginationLimit),
+	)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to create images pager")
+	}
+
+	for imagesPager.HasNext() {
+		images, err := imagesPager.GetNext()
+		if err != nil {
+			return "", errors.Wrap(err, "failed to get next images page")
+		}
+
+		for _, image := range images {
+			if *image.Name == imageName {
+
+				if *image.Status == "deprecated" {
+					l.Printf("WARN: image %s is deprecated, and will soon be unusable", *image.Name)
+				} else if *image.Status == "obsolete" {
+					return "", ErrImageObsolete
+				}
+
+				// Cache the image ID for future use.
+				p.cachedImageIDs[region][imageName] = *image.ID
+
+				return *image.ID, nil
+			}
+		}
+	}
+
+	return "", ErrImageNotFound
+}
+
+// getIdentityFromAuthenticator retrieves the identity from the authenticator.
+// This is a workaround for the fact that the authenticator does not
+// expose the identity directly. It uses the authenticator to create a
+// request, and then extracts the identity from the request's
+// Authorization header.
+// This is a temporary solution until, maybe one day, the authenticator is
```

Comment:
Quite a peculiar API it is... Curious if there is an external tracking issue for this?

Reply:
Didn't find any tracking.
In the end, this implementation is very similar with [what we had to do with Azure](https://github.com/cockroachdb/cockroach/blob/master/pkg/roachprod/vm/azure/auth.go#L90).